### DISCUSSION TITLE
Some useful extensions

### DISCUSSION THREAD

**Discussion by gurkantuna** [dotnet/csharplang#5219] [Language Ideas]
I'm using the simple extensions I've written below since a few weeks. Actually they only simplify to write check for `!= null` reference types or `>= 0` for value types, but even typing these long letters is tiring for me sometimes. Intellisense completing when I press the point button when using these extensions, although It is necessary to cast null controls between reference types bool and bool and is not useful in this case. So I'm open to ideas on how to get over this or objection.

```cs
public static class ExtensionObject {
    public static bool IsNull<T>(this T obj) where T : class => obj is null;
    public static bool IsNotNull<T>(this T obj) where T : class => obj is null;

    public static bool IsDefault<T>(this T value) where T : struct => value.Equals(default(T));
    public static bool IsNotDefault<T>(this T value) where T : struct => !value.Equals(default(T));
}

class Foo {
    public string PropAsStr { get; set; }
    public int PropAsInt { get; set; }
}
```

Below usage is useful for me.
```cs
var foo = new Foo();

if(foo.PropAsStr.IsNotNull()) {

}

if(foo.PropAsInt.IsDefault()) {

}

```
I want to use with `?` operator but compiler is waiting for cast from `bool?` to `bool`
```cs
if(foo?.PropAsStr.IsNotNull()) {

}

```

**Comment by svick** [comment]
I don't think these methods should be added to .Net (though that would be a request for dotnet/runtime, not this repo). They might make the code easier to type, but they also make it harder to read and I don't think that's a good trade-off.

I'm not opposed to making code like this easier to type, but I think it should be done purely on the IDE side. And [ReSharper already does this](https://www.jetbrains.com/help/resharper/Postfix_Templates.html), e.g. when you type `foo.PropAsStr.null`, it can generate the whole `null`-checking `if` statement for you.

**Reply by gurkantuna** [reply]
It's ok :) I actually was thinking this usage as an operator like ? operator or something else at least on the project level, but sure we don't have to agree :) Thanks for saying your opinion.

**Comment by PathogenDavid** [comment]
I feel like `foo.PropAsStr.IsNotNull()` is very odd compared to `foo.PropAsStr is not null`. (Also your `IsNotNull` implementation is wrong.)

If your goal is to optimizing typing, I feel like you'd be better off configuring snippets in your IDE.

----------

Back to your original question though...

The problem with using the null conditional operator here is that when the left side of the the `?` is null, the right side is not evaluated at all and the result of the expression is null. Since your expression would've resulted in a `bool` if the expression were to be evaluated this is `(bool?)null`.

In other words, your if statement is trying to compile down to this:

```csharp
if (foo is not null ? (bool?)foo.PropAsStr.IsNotNull() : (bool?)null)
```

Which isn't valid because you can't use `bool?` in an if statement's condition.

----------

If you're dead set on using your extension methods you either need to convert from a `bool?` to a `bool`, IE:

```csharp
if (foo?.PropAsStr.IsNotNull() ?? false)
```

or you need to change the order of operations so the `IsNotNull` applies to the entire expression, IE:

```csharp
if ((foo?.PropAsStr).IsNotNull())
```

**Comment by TonyValenti** [comment]
I have a lot of these extensions and more too.
